peer/brontide.go

```diff
@@ -2506,19 +2507,61 @@ func (p *Brontide) reenableActiveChannels() {
 	// both public and not pending.
 	activePublicChans := p.filterChannelsToEnable()
 
+	// Subscribe channel events early in case we miss events.
+	//
+	// NOTE: ChannelNotifier must be started before subscribing events
+	// otherwise we'd panic here.
+	sub, err := p.cfg.ChannelNotifier.SubscribeChannelEvents()
```

Comment:
I think this is a bit tricky, I'm pretty sure it's possible that:
- the link could have called `fail(...)`
- sent the `NotifyInactiveChannel` notification
- and _still_ be in the `activeChannels` map since the `handleLinkFailure` function may not have run yet in `peer/brontide.go`

This means that we would have missed the `NotifyInactiveChannel` notification and wait indefinitely for an event that won't come. Not entirely sure what to do here

Reply:
I'm thinking of maybe two possible solutions,
1. make the `onChannelFailure` blocking on `handleLinkFailure`, which means we need the method to signal back when it's finished. This way the `activeChannels` would be accurate.
2. make the subscription even earlier, maybe inside `addLink` so we won't miss any events.

I think an extension of solution 2 is to expand the usage of `ChannelNotifier` inside `peer` and we can craft actions based on the event, sort like the only communication channel between `link` and `peer`. Ofc this is just a thought, would require dramatic change I think.

Reply:
I think having two subsystems communicate with each other in the current way sucks because I've found and also introduced hard-to-debug deadlocks. So I don't like 1 unless we just change how we do it and NOT use go chans. Option 2 seems like a better idea because then we won't be able to miss any events, but also not perfect because we have to keep track of this notifier for each link.

If we go with option 2, there is one place in the peer that already subscribes to channel events (https://github.com/lightningnetwork/lnd/blob/6f944b13c8f7d139f6592f40a22e47deb94028ff/peer/brontide.go#L1298-L1310), so it may be worth it to have a dedicated `peer/brontide.go` goroutine to listen and pass the events to the other relevant peer goroutines.

Reply:
> so it may be worth it to have a dedicated peer/brontide.go goroutine to listen and pass the events to the other relevant peer goroutines.

So I tried to implement this as I like that a) we make early subscriptions to avoid missing events and b) no more one link one subscription, now all links share the same subscription.

But the issues are,
1) since each event is sent only once, what if the `ActiveLinkEvent` is wanted at both places?
2) if we decide to make the subscription early in the `(p *Brontide) Start()`, how do we deal with the old updates?

In the current approach, the subscription is canceled quickly, so I guess that's a pro. But it's prone to miss events, and that's a big con.

Reply:
For 1. if we had a dedicated goroutine in `peer/brontide.go`, we could duplicate the events and send it to the other goroutines that needed the event. Or we could just have multiple subscriptions per link - this option seems a bit easier.

For 2. not sure why old updates matter? Think we could ignore them?

Reply:
what is an old update here?

Reply:
a link can be activated and deactivated, then there would be a pair of events? By the time you want to check for `waitUntilLinkActive`, or any other places, how do you know if you are not receiving an old event?

Reply:
I don't think what you're describing can happen:

- `NotifyActiveLink` in `link.go` calls `SendUpdate`
  - The call to `SendUpdate` will block until the underlying `ntfnServer` is able to pull the update from the `s.updates` chan in `subscribe/subscribe.go`
  - When the update is pulled, the update is sent out to every client in the `s.clients` map.
- For a new peer object to be up that is also interested in this `ActiveLinkEvent`, the links must all be stopped beforehand. This is in `peerTerminationWatcher` in `server.go`.
  - If this new peer object were to receive an old event, it would need to already be in the `s.clients` map _at the same time_ that the update is pulled off the `s.updates` chan in `subscribe/subscribe.go`.
  - This isn't possible since the peer object would need to subscribe _before or at the same time_ that the `ActiveLinkEvent` went out. But the prior links must be stopped for this peer object to be started which is a contradiction